openvpn.8
.\"  OpenVPN -- An application to securely tunnel IP networks
.\"             over a single TCP/UDP port, with support for SSL/TLS-based
.\"             session authentication and key exchange,
.\"             packet encryption, packet authentication, and
.\"             packet compression.
.\"
.\"  Copyright (C) 2002-2003 James Yonan <jim@yonan.net>
.\"
.\"  This program is free software; you can redistribute it and/or modify
.\"  it under the terms of the GNU General Public License as published by
.\"  the Free Software Foundation; either version 2 of the License, or
.\"  (at your option) any later version.
.\"
.\"  This program is distributed in the hope that it will be useful,
.\"  but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"  GNU General Public License for more details.
.\"
.\"  You should have received a copy of the GNU General Public License
.\"  along with this program (see the file COPYING included with this
.\"  distribution); if not, write to the Free Software Foundation, Inc.,
.\"  59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
.\"
.\" Manual page for openvpn
.\" SH section heading
.\" SS subsection heading
.\" LP paragraph
.\" IP indented paragraph
.\" TP hanging label
openvpn
.TH openvpn 8 "29 October 2003"
.SH NAME
openvpn \- secure IP tunnel daemon.
.SH SYNOPSIS
.LP
.nh
.in +4
.ti -4
.B openvpn
[\ \fB\-\-help\fR\ ]
.in -4
.ti +4
.hy

.nh
.in +4
.ti -4
.B openvpn
[\ \fB\-\-config\fR\ \fIfile\fR\ ]
.in -4
.ti +4
.hy

.nh
.in +4
.ti -4
.B openvpn
[\ \fB\-\-genkey\fR\ ]
[\ \fB\-\-secret\fR\ \fIfile\fR\ ]
.in -4
.ti +4
.hy

.nh
.in +4
.ti -4
.B openvpn
[\ \fB\-\-mktun\fR\ ]
[\ \fB\-\-rmtun\fR\ ]
[\ \fB\-\-dev\fR\ \fItunX\ |\ tapX\fR\ ]
[\ \fB\-\-dev\-type\fR\ \fIdevice\-type\fR\ ]
[\ \fB\-\-dev\-node\fR\ \fInode\fR\ ]
.in -4
.ti +4
.hy

.nh
.in +4
.ti -4
.B openvpn
[\ \fB\-\-test\-crypto\fR\ ]
[\ \fB\-\-secret\fR\ \fIfile\fR\ ]
[\ \fB\-\-auth\fR\ \fIalg\fR\ ]
[\ \fB\-\-cipher\fR\ \fIalg\fR\ ]
[\ \fB\-\-keysize\fR\ \fIn\fR\ ]
[\ \fB\-\-no\-replay\fR\ ]
[\ \fB\-\-no\-iv\fR\ ]
.in -4
.ti +4
.hy

.nh
.in +4
.ti -4
.B openvpn
[\ \fB\-\-askpass\fR\ ]
[\ \fB\-\-auth\fR\ \fIalg\fR\ ]
[\ \fB\-\-ca\fR\ \fIfile\fR\ ]
[\ \fB\-\-cd\fR\ \fIdir\fR\ ]
[\ \fB\-\-cert\fR\ \fIfile\fR\ ]
[\ \fB\-\-chroot\fR\ \fIdir\fR\ ]
[\ \fB\-\-cipher\fR\ \fIalg\fR\ ]
[\ \fB\-\-comp\-lzo\fR\ ]
[\ \fB\-\-comp\-noadapt\fR\ ]
[\ \fB\-\-config\fR\ \fIfile\fR\ ]
[\ \fB\-\-crl\-verify\fR\ \fIcrl\fR\ ]
[\ \fB\-\-daemon\fR\ ]
[\ \fB\-\-dev\-name\fR\ \fIname\fR\ ]
[\ \fB\-\-dev\-node\fR\ \fInode\fR\ ]
[\ \fB\-\-dev\-type\fR\ \fIdevice\-type\fR\ ]
[\ \fB\-\-dev\fR\ \fItunX\ |\ tapX\ |\ null\fR\ ]
[\ \fB\-\-dh\fR\ \fIfile\fR\ ]
[\ \fB\-\-disable\-occ\fR\ ]
[\ \fB\-\-down\fR\ \fIcmd\fR\ ]
[\ \fB\-\-float\fR\ ]
[\ \fB\-\-fragment\fR\ \fImax\fR\ ]
[\ \fB\-\-gremlin\fR\ ]
[\ \fB\-\-group\fR\ \fIgroup\fR\ ]
[\ \fB\-\-hand\-window\fR\ \fIn\fR\ ]
[\ \fB\-\-ifconfig\fR\ \fIl\ r\fR\ ]
[\ \fB\-\-ifconfig\-noexec\fR\ ]
[\ \fB\-\-ifconfig\-nowarn\fR\ ]
[\ \fB\-\-inactive\fR\ \fIn\fR\ ]
[\ \fB\-\-inetd\fR\ ]
[\ \fB\-\-ipchange\fR\ \fIcmd\fR\ ]
[\ \fB\-\-key\fR\ \fIfile\fR\ ]
[\ \fB\-\-key\-method\fR\ \fIm\fR\ ]
[\ \fB\-\-keysize\fR\ \fIn\fR\ ]
[\ \fB\-\-link\-mtu\fR\ \fIn\fR\ ]
[\ \fB\-\-local\fR\ \fIhost\fR\ ]
[\ \fB\-\-log\-append\fR\ \fIfile\fR\ ]
[\ \fB\-\-log\fR\ \fIfile\fR\ ]
[\ \fB\-\-lport\fR\ \fIport\fR\ ]
[\ \fB\-\-mlock\fR\ ]
[\ \fB\-\-mssfix\fR\ \fImax\fR\ ]
[\ \fB\-\-mtu\-disc\fR\ \fItype\fR\ ]
[\ \fB\-\-mtu\-test\fR\ ]
[\ \fB\-\-mute\fR\ \fIn\fR\ ]
[\ \fB\-\-nice\-work\fR\ \fIn\fR\ ]
[\ \fB\-\-nice\fR\ \fIn\fR\ ]
[\ \fB\-\-no\-iv\fR\ ]
[\ \fB\-\-no\-replay\fR\ ]
[\ \fB\-\-nobind\fR\ ]
[\ \fB\-\-passtos\fR\ ]
[\ \fB\-\-persist\-key\fR\ ]
[\ \fB\-\-persist\-local\-ip\fR\ ]
[\ \fB\-\-persist\-remote\-ip\fR\ ]
[\ \fB\-\-persist\-tun\fR\ ]
[\ \fB\-\-ping\-exit\fR\ \fIn\fR\ ]
[\ \fB\-\-ping\-restart\fR\ \fIn\fR\ ]
[\ \fB\-\-ping\-timer\-rem\fR\ ]
[\ \fB\-\-ping\fR\ \fIn\fR\ ]
[\ \fB\-\-port\fR\ \fIport\fR\ ]
[\ \fB\-\-proto\fR\ \fIp\fR\ ]
[\ \fB\-\-remote\fR\ \fIhost\fR\ ]
[\ \fB\-\-reneg\-bytes\fR\ \fIn\fR\ ]
[\ \fB\-\-reneg\-pkts\fR\ \fIn\fR\ ]
[\ \fB\-\-reneg\-sec\fR\ \fIn\fR\ ]
[\ \fB\-\-replay\-persist\fR\ \fIfile\fR\ ]
[\ \fB\-\-replay\-window\fR\ \fIn\ [t]\fR\ ]
[\ \fB\-\-resolv\-retry\fR\ \fIn\fR\ ]
[\ \fB\-\-route\fR\ \fInetwork\ [netmask]\ [gateway]\ [metric]\fR\ ]
[\ \fB\-\-route\-delay\fR\ \fIn\fR\ ]
[\ \fB\-\-route\-gateway\fR\ \fIgw\fR\ ]
[\ \fB\-\-route\-noexec\fR\ ]
[\ \fB\-\-route\-up\fR\ \fIcmd\fR\ ]
[\ \fB\-\-rport\fR\ \fIport\fR\ ]
[\ \fB\-\-secret\fR\ \fIfile\fR\ ]
[\ \fB\-\-setenv\fR\ \fIname\ value\fR\ ]
[\ \fB\-\-shaper\fR\ \fIn\fR\ ]
[\ \fB\-\-single\-session\fR\ ]
[\ \fB\-\-tls\-auth\fR\ \fIf\fR\ ]
[\ \fB\-\-tls\-cipher\fR\ \fIl\fR\ ]
[\ \fB\-\-tls\-client\fR\ ]
[\ \fB\-\-tls\-remote\fR\ \fIx509name\fR\ ]
[\ \fB\-\-tls\-server\fR\ ]
[\ \fB\-\-tls\-timeout\fR\ \fIn\fR\ ]
[\ \fB\-\-tls\-verify\fR\ \fIcmd\fR\ ]
[\ \fB\-\-tran\-window\fR\ \fIn\fR\ ]
[\ \fB\-\-tun\-ipv6\fR\ ]
[\ \fB\-\-tun\-mtu\-extra\fR\ \fIn\fR\ ]
[\ \fB\-\-tun\-mtu\fR\ \fIn\fR\ ]
[\ \fB\-\-up\fR\ \fIcmd\fR\ ]
[\ \fB\-\-up\-restart\fR\ ]
[\ \fB\-\-up\-delay\fR\ ]
[\ \fB\-\-user\fR\ \fIuser\fR\ ]
[\ \fB\-\-verb\fR\ \fIn\fR\ ]
[\ \fB\-\-writepid\fR\ \fIfile\fR\ ]
.in -4
.ti +4
.hy
.SH DESCRIPTION
.LP
OpenVPN is a robust and highly configurable VPN (Virtual Private Network)
daemon which can be used to securely link two or more private networks
using an encrypted tunnel over the Internet. OpenVPN's principal strengths
include wide cross-platform portability, excellent stability,
support for dynamic IP addresses and NAT, adaptive link compression,
single TCP/UDP port usage, a modular design that offloads most crypto
tasks to the OpenSSL library, and relatively easy installation that in
most cases doesn't require a special kernel module.

OpenVPN is tightly bound to the OpenSSL library, and derives much
of its crypto capabilities from it.

OpenVPN supports
conventional encryption
using a pre-shared secret key
.B (Static Key mode)
or
public key security
.B (SSL/TLS mode)
using client & server certificates.
OpenVPN also
supports non-encrypted TCP/UDP tunnels.  

OpenVPN is designed to work with the
.B TUN/TAP
virtual networking interface that exists on most platforms.

Overall, OpenVPN aims to offer many of the key features of IPSec but
with a relatively lightweight footprint.
.SH OPTIONS
OpenVPN allows any option to be placed either on the command line
or in a configuration file.  Though all command line options are preceded
by a double-leading-dash ("--"), this prefix can be removed when
an option is placed in a configuration file.
.TP
.B --help
Show options.
.TP
.B --config file
Load additional config options from
.B file
where each line corresponds to one command line option,
but with the leading '--' removed.

Double quotation characters ("") can be used
to enclose single parameters containing whitespace,
and "#" or ";" characters in the first column
can be used to denote comments.

Configuration files can be nested to a reasonable depth.

For examples of configuration files,
see
.I http://openvpn.sourceforge.net/examples.html

Here is an example configuration file:
.RS
.ft 3
.nf
.sp
#
# Sample OpenVPN configuration file for
# using a pre-shared static key.
#
# '#' or ';' may be used to delimit comments.

# Use a dynamic tun device.
dev tun

# Our remote peer
remote mypeer.mydomain

# 10.1.0.1 is our local VPN endpoint
# 10.1.0.2 is our remote VPN endpoint
ifconfig 10.1.0.1 10.1.0.2

# Our pre-shared static key
secret static.key
.ft
.LP
.RE
.fi
.SS Tunnel Options:
.TP
.B --local host
Local host name or IP address.
If specified, OpenVPN will bind to this address only.
If unspecified, OpenVPN will bind to all interfaces.
.TP
.B --remote host
Remote host name or IP address.  If unspecified, OpenVPN will listen
for packets from any IP address, but will not act on those packets unless
they pass all authentication tests.  This requirement for authentication
is binding on all potential peers, even those from known and supposedly
trusted IP addresses (it is very easy to forge a source IP address on
a UDP packet).

When used in TCP mode, 
.B --remote
will act as a filter, rejecting connections from any host which does
not match
.B host.
.TP
.B --proto p
Use protocol
.B p
for communicating with remote host.
.B p
can be
.B udp,
.B tcp-client,
or
.B tcp-server.

The default protocol is
.B udp
when
.B --proto
is not specified.

For UDP operation,
.B --proto udp
should be specified on both peers.

For TCP operation, one peer must use
.B --proto tcp-server
and the other must use
.B --proto tcp-client.
A peer started with
.B tcp-server
will wait indefinitely for an incoming connection.  A peer
started with
.B tcp-client
will attempt to connect, and if that fails, will sleep for 5
seconds and try again.  Both TCP client and server will simulate
a SIGUSR1 restart signal if either side resets the connection.

OpenVPN is designed to operate optimally over UDP, but TCP capability is provided
for situations where UDP cannot be used.
In comparison with UDP, TCP will usually be
somewhat less efficient and less robust when used over unreliable or congested
networks.  TCP support is also a newer addition to OpenVPN, added in version
1.5, and should still be considered experimental until 1.6 or 1.7.

This article outlines some of problems with tunneling IP over TCP:

.I http://sites.inka.de/sites/bigred/devel/tcp-tcp.html

There are certain cases, however, where using TCP may be advantageous from
a security and robustness perspective, such as tunneling non-IP or
application-level UDP protocols, or tunneling protocols which don't
possess a built-in reliability layer.
.TP
.B --resolv-retry n
If hostname resolve fails for
.B --remote,
retry resolve for
.B n
seconds before failing (disabled by default).
.TP
.B --float
Allow remote peer to change its IP address and/or port number, such as due to
DHCP (this is the default if
.B --remote
is not used).
.B --float
when specified with
.B --remote
allows an OpenVPN session to initially connect to a peer
at a known address, however if packets arrive from a new
address and pass all authentication tests, the new address
will take control of the session.  This is useful when
you are connecting to a peer which holds a dynamic address
such as a dial-in user or DHCP client.

Essentially,
.B --float
tells OpenVPN to accept authenticated packets
from any address, not only the address which was specified in the
.B --remote
option.
.TP
.B --ipchange cmd
Execute shell command
.B cmd
when our remote ip-address is initially authenticated or
changes.

Execute as:

.B cmd ip_address port_number

See the "Environmental Variables" section below for
additional parameters passed as environmental variables.

Note that
.B cmd
can be a shell command with multiple arguments, in which
case all OpenVPN-generated arguments will be appended
to
.B cmd
to build a command line which will be passed to the script.

If you are running in a dynamic IP address environment where
the IP addresses of either peer could change without notice,
you can use this script, for example, to edit the
.I /etc/hosts
file with the current address of the peer.  The script will
be run every time the remote peer changes its IP address.

Similarly if
.I our
IP address changes due to DHCP, we should configure
our IP address change script (see man page for
.BR dhcpcd (8)
) to deliver a
.B SIGHUP
or
.B SIGUSR1
signal to OpenVPN.  OpenVPN will then
reestablish a connection with its most recently authenticated
peer on its new IP address.
.TP
.B --port port
TCP/UDP port number for both local and remote.
.TP
.B --lport port
TCP/UDP port number for local (default=5000).
.TP
.B --rport port
TCP/UDP port number for remote (default=5000).
.TP
.B --nobind
Do not bind to local address and port.  The IP stack will allocate
a dynamic port for returning packets.  Since the value of the dynamic port
could not be known in advance by a peer, this option is only suitable for
peers which will be initiating connections by using the
.B --remote
option.
.TP
.B --dev tunX | tapX | null
TUN/TAP virtual network device (
.B X
can be omitted for a dynamic device.)

See examples section below
for an example on setting up a TUN device.

You must use either tun devices on both ends of the connection
or tap devices on both ends.  You cannot mix them, as they
represent different underlying protocols.

.B tun
devices encapsulate IPv4 while
.B tap
devices encapsulate ethernet 802.3.
.TP
.B --dev-type device-type
Which device type are we using?
.B device-type
should be
.B tun
or
.B tap.
Use this option only if the TUN/TAP device used with
.B --dev
does not begin with
.B tun
or
.B tap.
.TP
.B --tun-ipv6
Build a tun link capable of forwarding IPv6 traffic.
Should be used in conjunction with
.B --dev tun
or
.B --dev tunX.
A warning will be displayed
if no specific IPv6 TUN support for your OS has been compiled into OpenVPN.
.TP
.B --dev-node node
Explicitly set the device node rather than using
/dev/net/tun, /dev/tun, /dev/tap, etc.  If OpenVPN
cannot figure out whether
.B node
is a TUN or TAP device based on the name, you should
also specify
.B --dev-type tun
or
.B --dev-type tap.

On Windows systems, select the TAP-Win32 adapter which
is named
.B node
in the Network Control Panel.  The
.B --show-adapters
option under Windows can also be used
to enumerate all available TAP-Win32
adapters.
.TP
.B --ifconfig l rn
Set TUN/TAP adapter parameters. 
.B l
is the IP address of the local VPN endpoint.
For TUN devices,
.B rn
is the IP address of the remote VPN endpoint.
For TAP devices,
.B rn
is the subnet mask of the virtual ethernet segment
which is being created or connected to.

For TUN devices, which facilitate virtual
point-to-point IP connections,
the proper usage of
.B --ifconfig
is to use two private IP addresses
which are not a member of any
existing subnet which is in use.
The IP addresses may be consecutive
and should have their order reversed
on the remote peer.  After the VPN
is established, by pinging
.B rn,
you will be pinging across the VPN.

For TAP devices, which provide
the ability to create virtual
ethernet segments,
.B --ifconfig
is used to set an IP address and
subnet mask just as a physical
ethernet adapter would be
similarly configured.  If you are
attempting to connect to a remote
ethernet bridge, the IP address
and subnet should be set to values
which would be valid on the
the bridged ethernet segment (note
also that DHCP can be used for the
same purpose).

This option, while primarily a proxy for the
.BR ifconfig (8)
command, is designed to simplify TUN/TAP
tunnel configuration by providing a
standard interface to the different
ifconfig implementations on different
platforms.

.B --ifconfig
parameters which are IP addresses can
also be specified as a DNS or /etc/hosts
file resolvable name.
.TP
.B --ifconfig-noexec
Don't actually execute ifconfig/netsh commands, instead
pass
.B --ifconfig
parameters to scripts using environmental variables.
.TP
.B --ifconfig-nowarn
Don't output an options consistency check warning
if the
.B --ifconfig
option on this side of the
connection doesn't match the remote side.  This is useful
when you want to retain the overall benefits of the
options consistency check (also see
.B --disable-occ
option) while only disabling the ifconfig component of
the check.

For example,
if you have a configuration where the local host uses
.B --ifconfig
but the remote host does not, use
.B --ifconfig-nowarn
on the local host.
.TP
.B --route network [netmask] [gateway] [metric]
Add route to routing table after connection is established.
Multiple routes can be specified.  Routes will be
automatically torn down in reverse order prior to
tun/tap device close.

This option is intended as
a convenience proxy for the
.BR route (8)
shell command,
while at the same time providing portable semantics
across OpenVPN's platform space.

.B netmask
default -- 255.255.255.255

.B gateway
default -- taken from
.B --route-gateway
or the second parameter to
.B --ifconfig
when
.B --dev tun
is specified.

The default can be specified by leaving an option blank or setting
it to "default".

The
.B network
and
.B gateway
parameters can
also be specified as a DNS or /etc/hosts
file resolvable name.
.TP
.B --route-gateway gw
Specify a default gateway
.B gw
for use with
.B --route.
.TP
.B --route-delay [n]
Delay
.B n
seconds (default=0) after connection
establishment, before adding routes. If
.B n
is 0, routes will be added immediately upon connection
establishment.  If
.B --route-delay
is omitted, routes will be added immediately after tun/tap device
open and
.B --up
script execution, before any
.B --user
or 
.B --group
privilege downgrade (or
.B --chroot
execution.)

This option is designed to be useful in scenarios where DHCP is
used to set
tap adapter addresses.  The delay will give the DHCP handshake
time to complete before routes are added.

A better solution would be to poll the virtual adapter, waiting
for its IP address to become defined, but unfortunately there
is no platform independent mechanism for doing this.
.TP
.B --route-up cmd
Execute shell command
.B cmd
after routes are added, subject to
.B --route-delay.

See the "Environmental Variables" section below for
additional parameters passed as environmental variables.

Note that
.B cmd
can be a shell command with multiple arguments.
.TP
.B --route-noexec
Don't add or remove routes automatically.  Instead pass routes to
.B --route-up
script using environmental variables.
.TP
.B --link-mtu n
Take the link device MTU to be n and derive the TUN MTU
from it (default=1300 for TUN devices).  The
default is a conservative
value that was chosen
because it has a higher probability of working correctly.
However, for many
cases, using a value of 1472 will maximize performance for TUN devices
over IPv4.

Fundamentally,
.B --link-mtu
sets an upper bound on the size of UDP packets which are sent
between OpenVPN peers.

Prior to OpenVPN 1.5, this option was named
.B --udp-mtu.
While deprecated, this name will still be supported
for compatibility.

This option should only be used for TUN style tunnels.  TAP tunnels
that are used for ethernet bridging should use the
.B --tun-mtu
parameter.

The MTU (Maximum Transmission Units) is
the maximum datagram size in bytes that can be sent unfragmented
over a particular network path.  OpenVPN requires that packets
on the control or data channels be sent unfragmented.

Typically, the link MTU should be set to a value between 1300 and 1500.
The optimal size for link MTU is the largest
MTU that can be handled by every router on the link path.
The link MTU value should be equal on both peers.

MTU problems often manifest themselves as connections which
hang during periods of active usage.  The
.B --fragment
and
.B --mssfix
options are provided as workarounds to such problems.

OpenVPN
adds a small amount of overhead to each tunnel packet before
it is forwarded from the TUN device over the secure link channel.
This overhead consists of data fields such as the HMAC signature,
packet ID, encryption block padding, etc.  Because of this overhead,
the TUN device MTU should be slightly smaller than the link device
MTU to make room for the extra bytes which OpenVPN adds to every
data channel packet.  OpenVPN allows you to explicitly specify either
the TUN MTU or the link MTU (but not both).  OpenVPN will then
compute the value you didn't specify based on the value you did.
OpenVPN will compute exactly how much overhead it will need to add
to each packet, based on the other options you specify.  If you
specify an
.B --up
script, OpenVPN will pass the TUN MTU and link MTU values on the command line
to the script.
.TP
.B --tun-mtu n
Take the TUN device MTU to be
.B n
and derive the link MTU
from it (default=1500 for TAP devices).

See
.B --link-mtu
above more more information on MTU.
.TP
.B --tun-mtu-extra n
Assume that the TUN/TAP device might return as many as
.B n
bytes more than the
.B --tun-mtu
size on read.  This parameter defaults to 0, which is sufficient for
most TUN devices.  TAP devices may introduce additional overhead in excess
of the MTU size, and a setting of 32 is the default when TAP devices are used.
This parameter only controls internal OpenVPN buffer sizing,
so there is no transmission overhead associated with using a larger value.
.TP
.B --mtu-disc type
Should we do Path MTU discovery on TCP/UDP channel?  Only supported on OSes such
as Linux that supports the necessary system call to set.

.B 'no'
-- Never send DF (Don't Fragment) frames
.br
.B 'maybe'
-- Use per-route hints
.br
.B 'yes'
-- Always DF (Don't Fragment)
.br
.TP
.B --mtu-test
To empirically measure MTU on connection startup,
add the
.B --mtu-test
option to your configuration.
OpenVPN will send ping packets of various sizes
to the remote peer and measure the largest packets
which were successfully received.  The
.B --mtu-test
process normally takes about 3 minutes to complete.
.TP
.B --fragment max
Enable internal datagram fragmentation so
that no UDP datagrams are sent which
are larger than
.B max
bytes.

The
.B max
parameter is interpreted in the same way as the
.B --link-mtu
parameter, i.e. the UDP packet size after encapsulation
overhead has been added in, but not including
the UDP header itself.

The
.B --fragment
option only makes sense when you are using the UDP protocol
for OpenVPN peer-to-peer communication, i.e.
.B --proto udp.

.B --fragment
adds 4 bytes of overhead per datagram.

See the
.B --mssfix
option below for an important companion option to
.B --fragment.

It should also be noted that this option is not meant to replace
UDP fragmentation at the IP stack level.  It is only meant as a
last resort when path MTU discovery is broken.  Using this option
is less efficient than fixing path MTU discovery for your IP link and
using native IP fragmentation instead.

Having said that, there are circumstances where using OpenVPN's
internal fragmentation capability may be your only option, such
as tunneling a UDP multicast stream which requires fragmentation.
.TP
.B --mssfix max
Announce to TCP sessions running over the tunnel that they should limit
their send packet sizes such that after OpenVPN has encapsulated them,
the resulting UDP packet size that OpenVPN sends to its peer will not
exceed
.B max
bytes.

The
.B max
parameter is interpreted in the same way as the
.B --link-mtu
parameter, i.e. the UDP packet size after encapsulation
overhead has been added in, but not including
the UDP header itself.

The
.B --mssfix
option only makes sense when you are using the UDP protocol
for OpenVPN peer-to-peer communication, i.e.
.B --proto udp.


ChangeLog
OpenVPN
Copyright (C) 2002-2004 James Yonan <jim@yonan.net>

$Id: ChangeLog,v 1.89.2.6 2004/03/02 22:02:03 jimyonan Exp $

2004.03.02 -- Version 2.0-test6

* Branch merge with 1.6-rc1.

2004.03.02 -- Version 2.0-test5

* Move Socks5 UDP header append/remove to socks.c, and is
  called from forward.c.
* Moved verify statics from ssl.c into struct tls_session.
* Wrote multi.[ch] to handle top level of point-to-multipoint
  mode.
* Wrote some code to allow a struct link_socket in a child context
  to be slaved to the parent context.
* Broke up packet read and process functions in forward.c
  (from socket or tuntap) into separate functions for read
  and process, so that point-to-point and point-to-multipoint can
  share the same code.
* Expand TLS control channel to allow the passing of configuration
  commands.
* Wrote mroute.[ch] to handle internal packet routing for
  point-to-multipoint mode.

2004.02.22 -- Version 2.0-test3

* Initial work on UDP multi-client server.
* Branch merge of 1.6-beta7
	
2004.02.14 -- Version 2.0-test2

* Refactorization of openvpn.c into openvpn.[ch]
  init.[ch] forward.[ch] forward-inline.h
  occ.[ch] occ-inline.h  ping.[ch] ping-inline.h
  sig.[ch].  Created a master per-tunnel
  struct context in openvpn.h.
* Branch merge of 1.6-beta6.2

2003.11.06 -- Version 2.0-test1

* Initial testbed for 2.0.

2004.02.22 -- Version 1.6-beta7
	
* Allow scheduling priority increase (--nice) together
  with UID/GID downgrade (--user/--group).
* Code that causes SIGUSR1 restart on TLS errors in TCP
  mode was not activated in pthread builds.
* Save the certificate serial number in an environmental
  variable called tls_serial_{n} prior to calling the
  --tls-verify script.  n is the current cert chain level.
* Added NetBSD IPv6 tunnel capability (also requires
  a kernel patch) (Horst Laschinsky).
* Fixed bug in checking the return value of the nice()
  function (Ian Pilcher).
* Bug fix in new FreeBSD IPv6 over TUN code which was
  originally added in 1.6-beta5 (Nathanael Rensen).
* More Socks5 fixes -- extended the struct frame
  infrastructure to accomodate proxy-based encapsulation
  overhead.
* Added --dhcp-option to Windows version for setting
  adapter properties such as WINS & DNS servers.
* Use a default route-delay of 5 seconds when
  --ip-win32 dynamic is specified (only applicable when
  --route-delay is not explicitly specified).
* Added "log_append" registry variable to control
  whether the OpenVPN service wrapper on Windows
  opens log files in append (log_append="1") or
  truncate (log_append="0") mode.  The default
  is truncate.

2004.02.05 -- Version 1.6-beta6

* UDP over Socks5 fix to accomodate Socks5 encapsulation
  overhead (Christof Meerwald).
* Minor --ip-win32 dynamic tweaks (use long lease time,
  invalidate existing lease with DHCPNAK).

2004.02.01 -- Version 1.6-beta5

* Added Socks5 proxy support (Christof Meerwald).
* IPv6 tun support for FreeBSD (Thomas Glanzmann).
* Special TAP-Win32 debug mode for Windows self-install that was
  enabled in beta4 is now turned off.
* Added some new Solaris notes to INSTALL (Koen Maris).
* More work on --ip-win32 dynamic.

2004.01.27 -- Version 1.6-beta4

* For this beta, the Windows self-install is a debug version
  and will run slower -- use only for testing.
* Reverted the --ip-win32 default back to 'ipapi'
  from 'dynamic'.
* Added the offset parameter to '--ip-win32 dynamic' which
  can be used to control the address of the masqueraded
  DHCP server which replies to Windows DHCP requests.
* Added a wait/nowait option to --inetd (nowait can only
  be used with TCP sockets, TLS authentication, and over
  a bridged configuration -- see FAQ for more info)
  (Stefan `Sec` Zehl).
* Added a build-time capability where TAP-Win32 driver
  debug messages can be output by OpenVPN at --verb 6
  or higher.

2004.01.20 -- Version 1.6-beta2

* Added ./configure --enable-iproute2 flag which
  uses iproute2 instead of route + ifconfig --
  this is necessary for the LEAF Linux distro
  (Martin Hejl).
* Added renewal-time and rebind-time to set of
  DHCP options returned by the TAP-Win32 driver when
  "--ip-win32 dynamic" is used.
	
2004.01.14 -- Version 1.6-beta1

* Fixed --proxy bug that sometimes caused plaintext
  control info generated by the proxy prior to http
  CONNECT method establishment to be incorrectly
  parsed as OpenVPN data.
* For Windows version, implemented the
  "--ip-win32 dynamic" method and made it the default.
  This method sets the TAP-Win32 adapter IP address
  and netmask by replying to the kernel's DHCP queries.
  See the man page for more detailed info.
* Added --connect-retry parameter which controls
  the time interval (in seconds) between connect()
  retries when --proto tcp-client is used.  Previously,
  this value was hardcoded to 5 seconds, and still
  defaults as such.
* --resolv-retry can now be used with a parameter
  of "infinite" to retry indefinitely.
* Added SSL_CTX_use_certificate_chain_file() to ssl.c
  for support of multi-level certificate chains
  (Sten Kalenda).
* Fixed --tls-auth incompatibility with 1.4.x and earlier
  versions of OpenVPN when the passphrase file is an
  OpenVPN static key file (as generated by --genkey).
* Added shell-escape support in config files using
  the backslash character ("\") so that (for example)
  double quotes can be passed to the shell.
* Added "contrib" subdirectory on tarball, source zip,
  and CVS containing user-submitted contributions.
* Added an optional patch to the Redhat init script to
  allow the configuration file directory to be a
  multi-level directory hierarchy (Farkas Levente).
  See contrib/multilevel-init.patch
* Added some scripts and documentation on using
  Linux "fwmark" iptables rules to enable
  fine-grained routing control over the VPN
  (Sean Reifschneider, <jafo@tummy.com>).
  See contrib/openvpn-fwmarkroute-1.00

2003.11.20 -- Version 1.5.0

* Minor documentation changes.

2003.11.04 -- Version 1.5-beta14


openvpn.c
/*
 *  OpenVPN -- An application to securely tunnel IP networks
 *             over a single TCP/UDP port, with support for SSL/TLS-based
 *             session authentication and key exchange,
 *             packet encryption, packet authentication, and
 *             packet compression.
 *
 *  Copyright (C) 2002-2004 James Yonan <jim@yonan.net>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program (see the file COPYING included with this
 *  distribution); if not, write to the Free Software Foundation, Inc.,
 *  59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#ifdef WIN32
#include "config-win32.h"
#else
#include "config.h"
#endif

#include "syshead.h"

#include "init.h"
#include "forward.h"

#include "memdbg.h"

int
main (int argc, char *argv[])
{
  struct context c;
  const int gc_level = gc_new_level ();

  /* signify first time for components which can
     only be initialized once per program instantiation. */
  c.first_time = true;

  /* initialize program-wide statics */
  init_static ();

  /*
   * This loop is initially executed on startup and then
   * once per SIGHUP.
   */
  do
    {
      context_clear_all_except_first_time (&c);

      /* static signal info object */
      c.sig = &siginfo_static;

      /* initialize options to default state */
      init_options (&c.options);

      /* parse command line options, and read configuration file */
      parse_argv (&c.options, argc, argv);

      /* init verbosity and mute levels */
      init_verb_mute (&c.options);

      /* set dev options */
      init_options_dev (&c.options);

      /* openssl print info? */
      if (print_openssl_info (&c.options))
	break;

      /* --genkey mode? */
      if (do_genkey (&c.options))
	break;

      /* tun/tap persist command? */
      if (do_persist_tuntap (&c.options))
	break;

      /* sanity check on options */
      options_postprocess (&c.options, c.first_time);

      /* test crypto? */
      if (do_test_crypto (&c.options))
	break;

      /* misc stuff */
      pre_setup (&c.options);

      /* finish context init */
      context_init_1 (&c);

      do
	{
	  const int gc_level_inner = gc_new_level ();

	  context_clear_2 (&c);

	  /* initialize tunnel instance */
	  init_instance (&c);

	  /* main event loop */
	  while (true)
	    {
	      /* process timers, TLS, etc. */
	      pre_select (&c);

	      /* garbage collect */
	      gc_collect (gc_level_inner);

	      /* set up and do the select() */
	      single_select (&c);

	      /* process signals */
	      if (c.sig->signal_received)
		{
		  if (c.sig->signal_received == SIGUSR2)
		    {
		      print_status (&c);
		      c.sig->signal_received = 0;
		      continue;
		    }
		  print_signal (c.sig->signal_received);
		  break;
		}

	      /* timeout? */
	      if (!c.c2.select_status)
		continue;

	      /* process the I/O which triggered select */
	      process_io (&c);

	      if (c.sig->signal_received)
		break;
	    }

	  /* tear down tunnel instance (unless --persist-tun) */
	  close_instance (&c);
	  c.first_time = false;
	  gc_free_level (gc_level_inner);

	}
      while (c.sig->signal_received == SIGUSR1);

      gc_collect (gc_level);

    }
  while (c.sig->signal_received == SIGHUP);

  /* uninitialize program-wide statics */
  uninit_static ();

  /* pop our garbage collection level */
  gc_free_level (gc_level);

  openvpn_exit (OPENVPN_EXIT_STATUS_GOOD);	/* exit point */
  return 0;			/* NOTREACHED */
}
